### ISSUE TITLE
Conversion from JSON

### ISSUE THREAD

**Issue by johnjcamilleri** [GrammaticalFramework/gf-typescript#1]
I started writing a function `fromJSON` which would convert a GF grammar compiled to canonical JSON into an object of the `GFGrammar` class, thinking this would eventually replace the need for the JS grammar format.

However on closer analysis, this is not trivial at all. Below I have posted a sample grammar, with the output as both JavaScript (with `gf --make --output-format=js`) and JSON (with `gf --make --output-format=canonical_json`). Immediately one notices that the parameters are preserved in the JSON format but not the JS. After confirming with the GF compiler source code, I realise that the JSON output is a representation of the GF source while the JS output is actually a representation of the compiled PGF.

This leaves me with the following options/questions. I'm tagging @heatherleaf @krangelov @Thomas-H who might have some valuable input here.

1. To proceed with the current formats would mean I need to write a Canonical-JSON to PGF-JS converter, i.e. re-implement the GF compiler in JavaScript! 😣 This seems unreasonable, unless I've missed something which means the task is actually not that complicated.
2. We could introduce a PGF-JSON format, essentially very similar to PGF-JS (and could replace it) but which might be useful in other contexts. I could imagine writing this myself, based on `gf-core:src/compiler/GF/Compile/PGFtoJS.hs`, if deemed valuable.
3. Since the PGF-JS format is closer to PGF than GF, how safe is it use publicly (if you consider your grammars to be private)? Can one obtain a reasonable GF grammar backwards from its compiled PGF representation, or is so much information compiled away that this would serve as effective protection against obtaining the original grammar? (this pertains to a discussion I had elsewhere with @krangelov).

**Zero.gf**
```gf
abstract Zero = {
  flags
    startcat = Utt ;

  cat
    Utt ;
    N ;

  fun
    eat: N -> Utt ;
    apple, banana: N ;
}
```

**ZeroSwe.gf**
```gf
concrete ZeroSwe of Zero = {
  param
    Gender = Utrum | Neutrum ;

  lincat
    Utt = { s: Str } ;
    N = { s: Str ; g: Gender } ;

  lin
    eat n = { s = "äta" ++ artIndef ! n.g ++ n.s } ;

    apple = { s = "äpple" ; g = Neutrum } ;
    banana = { s = "banan" ; g = Utrum } ;

  oper
    artIndef : Gender => Str = table {
      Utrum => "en" ;
      Neutrum => "ett"
    } ;
}
```

**Zero.js**
```js
var Zero = new GFGrammar(new GFAbstract('Utt', {
  apple: new Type([], 'N'),
  banana: new Type([], 'N'),
  eat: new Type(['N'], 'Utt')
}), {
  ZeroSwe: new GFConcrete({}, {
    0: [new Apply(5, [])],
    1: [new Apply(4, [])],
    2: [new Apply(6, [new PArg(0)]), new Apply(7, [new PArg(1)])]
  }, [
    new CncFun('\'lindef N\'', [1]),
    new CncFun('\'lindef N\'', [0]),
    new CncFun('\'lindef Utt\'', [1]),
    new CncFun('\'lindef Utt\'', [0]),
    new CncFun('apple', [3]),
    new CncFun('banana', [2]),
    new CncFun('eat', [4]),
    new CncFun('eat', [5])
  ], [
    [new SymCat(0, 0)],
    [new SymLit(0, 0)],
    [new SymKS('banan')],
    [new SymKS('äpple')],
    [new SymKS('äta'), new SymKS('en'), new SymCat(0, 0)],
    [new SymKS('äta'), new SymKS('ett'), new SymCat(0, 0)]
  ], {
    Float: {
      s: -3,
      e: -3
    },
    Int: {
      s: -2,
      e: -2
    },
    N: {
      s: 0,
      e: 1
    },
    String: {
      s: -1,
      e: -1
    },
    Utt: {
      s: 2,
      e: 2
    }
  }, 3)
})
```

**Zero.json**
```json
{
  "abstract": {
    "abs": "Zero",
    "flags": {
      "startcat": "Utt"
    },
    "cats": ["N", "Utt"],
    "funs": [{
      "fun": "apple",
      "type": {
        ".args": [],
        ".result": "N"
      }
    }, {
      "fun": "banana",
      "type": {
        ".args": [],
        ".result": "N"
      }
    }, {
      "fun": "eat",
      "type": {
        ".args": ["N"],
        ".result": "Utt"
      }
    }]
  },
  "concretes": [{
    "cnc": "ZeroSwe",
    "abs": "Zero",
    "flags": {},
    "params": [{
      "param": "ZeroSwe.Gender",
      "values": ["ZeroSwe.Utrum", "ZeroSwe.Neutrum"]
    }],
    "lincats": [{
      "cat": "N",
      "lintype": {
        "s": "Str",
        "g": "ZeroSwe.Gender"
      }
    }, {
      "cat": "Utt",
      "lintype": {
        "s": "Str"
      }
    }],
    "lins": [{
      "fun": "apple",
      "args": [],
      "lin": {
        "s": "äpple",
        "g": {
          ".param": "ZeroSwe.Neutrum"
        }
      }
    }, {
      "fun": "banana",
      "args": [],
      "lin": {
        "s": "banan",
        "g": {
          ".param": "ZeroSwe.Utrum"
        }
      }
    }, {
      "fun": "eat",
      "args": ["N_0"],
      "lin": {
        "s": ["äta", {
          ".select": {
            ".tblarg": "ZeroSwe.Gender",
            ".tblrows": [{
              ".pattern": "ZeroSwe.Utrum",
              ".value": "en"
            }, {
              ".pattern": "ZeroSwe.Neutrum",
              ".value": "ett"
            }]
          },
          ".key": {
            ".project": {
              ".var": "N_0"
            },
            ".label": "g"
          }
        }, {
          ".project": {
            ".var": "N_0"
          },
          ".label": "s"
        }]
      }
    }]
  }]
}
```

**Comment by krangelov** [comment]
You can definitely decompile a PGF file back to GF source. However the
source code would not be very readable. I guess this counts as obfuscation.

Otherwise, option 2 seems the most promissing. It would also eliminate the
need for similar existing exports to Prolog and Python.




На сб, 29.06.2019 г., 13:50 John J. Camilleri <notifications@github.com>
написа:

**Comment by heatherleaf** [comment]
> 1. To proceed with the current formats would mean I need to write a Canonical-JSON to PGF-JS converter, i.e. re-implement the GF compiler in JavaScript! 😣 This seems unreasonable, unless I've missed something which means the task is actually not that complicated.

No, you didn't miss anything. The canonical format is before compiling to PMCFG, and its main purpose is to be able to do automatic grammar conversions (e.g., to conflate parameters: https://github.com/MUSTE-Project/gf-canonical-transforms)

> 2. We could introduce a PGF-JSON format, essentially very similar to PGF-JS (and could replace it) but which might be useful in other contexts. I could imagine writing this myself, based on `gf-core:src/compiler/GF/Compile/PGFtoJS.hs`, if deemed valuable.

This is what I suggest, and as @krangelov writes could hopefully eliminate the need for other PGF/PMCFG exports.

> 3. Since the PGF-JS format is closer to PGF than GF, how safe is it use publicly (if you consider your grammars to be private)? Can one obtain a reasonable GF grammar backwards from its compiled PGF representation, or is so much information compiled away that this would serve as effective protection against obtaining the original grammar? (this pertains to a discussion I had elsewhere with @krangelov).

As @krangelov writes, it's not difficult to create a GF from a PGF, but it's not the original GF source. The new GF file will not have any parameters or tables - am I right, @krangelov ?

**Comment by johnjcamilleri** [comment]
Ok, I have started working on the JSON export in the [gf-core/pgf2json](https://github.com/GrammaticalFramework/gf-core/tree/pgf2json) branch.
I will return to this issue once that's done.

**Comment by johnjcamilleri** [comment]
Closed by 82cd9edd1d9f80bc0baa26fa50528dbd521aeb38
